### ISSUE TITLE
Query on PySINDy - Adoption into current developing model to fit the experimental data

### ISSUE THREAD

**Issue by CKhyd** [dynamicslab/pysindy#314]
I am developing a water quality model and very interested in SINDy.
I have developed a water quality model assuming a closed system called below a.

```
def a(t,X):   
    rLPOC    =  0.12/24
    rLDOC    =  0.25/24
    rLPON    =  0.12/24
    rLDON    =  0.25/24
    rhT      =  0.069
    TNit     =  27 
    rNitM    =  0.05/24
    rKhNitDO =  4
    rKhNitN  =  0.2
    rKnit1   =  0.0045
    rKnit2   =  0.0045
    ANDC     =  0.933
    AANOX    =  0.5
    rKHORDO  =  1
    rKHDNn   =  0.1
    AOC      =  2.67
    AON      =  4.33
    temp     =  20
    
    LPOC_1,LDOC_1,DIC_1,LPON_1,LDON_1,NH4_1,NO23_1,N2_1,DO_1  =  X
    
    # LPOC
    hyd_c  = rLPOC*np.exp(rhT*(temp-20))
    dLPOCdt= -LPOC_1*hyd_c
    
    # LDOC
    mine_c = rLDOC*np.exp(rhT*(temp-20))*DO_1/(DO_1+rKHORDO)                      
    denit  = rKHORDO*NO23_1*AANOX*mine_c/((rKHORDO+DO_1)*(rKHDNn+NO23_1))
    dLDOCdt= LPOC_1*hyd_c -LDOC_1*mine_c -LDOC_1*denit
    
    #DIC
    dDICdt = LDOC_1*mine_c+LDOC_1*denit
    
    # LPON
    hyd_n  = rLPON*np.exp(rhT*(temp-20))                                  
    dLPONdt= -LPON_1*hyd_n
    
    # LDON
    mine_n = rLDON*np.exp(rhT*(temp-20))        
    dLDONdt= LPON_1*hyd_n -LDON_1*mine_n

    # NH4
    if temp>=TNit:
        tmp = rKnit1 
    else:
        tmp = rKnit2 
    f      = np.exp(-1*tmp*((temp-TNit)**2))
    nit    = rNitM*(DO_1/(rKhNitN+DO_1))*(NH4_1/(rKhNitN+NH4_1))*f
    dNH4dt = LDON_1*mine_n-NH4_1*nit
    
    #NO23
    dNO23dt= NH4_1*nit    -ANDC*denit*LDOC_1
    
    #N2
    dN2dt  = ANDC*denit*LDOC_1 
    
    #DO
    dDOdt  = -AON*nit*NH4_1-AOC*mine_c*LDOC_1
    
    return [dLPOCdt,dLDOCdt,dDICdt,dLPONdt,dLDONdt,dNH4dt,dNO23dt,dN2dt,dDOdt]      
```

However, the experimental data (DO, Hour_DO) and the model's simulation results differed significantly.

```
ntime = 60*24
init = [0.9, 0.5, 0, 0.21, 0.14, 0.4, 0.01, 0, 7.0]
t_span = [0.0,ntime-1]
t_eval = np.linspace(*t_span,ntime)      
sol = integrate.solve_ivp(a,t_span,init,t_eval=t_eval)
DO_t   = sol.y[8]

fig = plt.figure(figsize=(4, 4))
ax1 = fig.add_subplot(1, 1, 1)

ax1.plot(np.linspace(0,ntime-1,ntime),DO_t,'k',lw=1)
ax1.scatter(Hour_DO,DO)
ax1.set_axisbelow(True)
ax1.set_title('DO',fontsize=20)
ax1.set_xlim([0,60*24])
ax1.set_ylim([0,10])

```

My question is, using PySINDy, is it possible to change only the ODE of dDOdt in a so that it fits the experimental data (DO, Hour_DO)?
I may be asking the wrong question due to my inexperience, but I would appreciate your response.

**Comment by CKhyd** [comment]
```
DO = np.array([7.,6.,5.5,5.25,4.5,4.,3.5,3.,2.5,2.1,2.,1.5,1.4,1.3,1.1])
Hour_DO = np.array([ 0,4,14,18,36,48,60,84,108,132,228,300,348,396,480])
```

**Comment by Jacob-Stevens-Haas** [comment]
Yes, that is the purpose of SINDy, but you don't have a ton of data here.  In particular, if you believe the dynamics involve nine variables (`LPOC_1,LDOC_1,DIC_1,LPON_1,LDON_1,NH4_1,NO23_1,N2_1,DO_1`), you might need to measure more of those variables than just the two (`DO` and `Hour_DO`).  You don't necessarily need all, but probably need most.

**Comment by CKhyd** [comment]
Thank you for your response.
All the mentioned variables are available. 

However, may I know if it possible to change only the ODE of dDOdt of a without changing the others? 
I have prepared some provisional experimental data.

[obs.csv](https://github.com/dynamicslab/pysindy/files/11280535/obs.csv)

From left to right: time, LPOC_1, LDOC_1, DIC_1, LPON_1, LDON_1, NH4_1, NO23_1, N2_1, DO_1, and I added temp.
Is it possible to change only the ODE of dDOdt of a from these data?

I would appreciate it if you could let me know, as I have no idea how to write the code.
Thanks in advance for any help you can give me.

**Comment by Jacob-Stevens-Haas** [comment]
tl;dr - no, it's not possible.  `pysindy.SINDy` models do not have the ability to freeze components, nor generally to retrain on new data.  In other words, running 
```python
model.fit(x1, t1)
model.fit(x2, t2)
```
is in most cases the same as
```python
model.fit(x2, t2)
```

However, this is not guaranteed.  From [`scikit-learn.BaseEstimator`](https://scikit-learn.org/stable/developers/develop.html#fitting):

>When fit is called, any previous call to fit should be ignored. In general, calling estimator.fit(X1) and then estimator.fit(X2) should be the same as only calling estimator.fit(X2). However, this may not be true in practice when fit depends on some random process, see [random_state](https://scikit-learn.org/stable/glossary.html#term-random_state). Another exception to this rule is when the hyper-parameter warm_start is set to True for estimators that support it. warm_start=True means that the previous state of the trainable parameters of the estimator are reused instead of using the default initialization strategy.
